fix(css): elaborate how gap percentage values behave (#34018)
mdn/content: files/en-us/web/css/gap/index.md

{{EmbedLiveSample("Multi-column_layout", "auto", "120px")}}

## Percentage gap values

Gap acts like an empty item/track. In other words instead of gap you could simply use an empty item/track in the container. The item/track's size would be the size set in `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.

### Explicit container size

If the container has fixed size set then percentage value calculations are simple. Percentages are simply calculated from the sizes set on the container. Thus gap behaviour is consistant across all layouts. In the following screenshot there are two containers one with grid layout and other with flex layout. The containers have five red 20x20px children. Both the containers have height explicitly set to 200px using `height: 200px`. And the gap is set using `gap: 10% 0`.

> ![Screenshot demoing gap sizes in grid and flex layouts with fixed 200px height set.](fixed_height.jpg)

Reviewer: Quick one before I go deeper in this review: shall we do a live sample instead of screenshots?

Author: The screenshots can't be replicated using livesample; it is not possible to color gaps. In developer tools when you select grid/flex elements in HTML tab you get to see the sizes, all the gaps, and the highlighting.

Reviewer: We could do live samples and point readers where to look for layout inspector in developer tools to highlight the gaps? I would really like to avoid adding more screenshots 😂 and the code can be remixed and maintained afterwards! What do you think?

Author: removed the screenshots and used live samples instead

Reviewer: Looks much better IMO, thank you! 🙏🏻

{{EmbedLiveSample("Multi-column_layout", "auto", "120px")}}

> ## Percentage gap values

Reviewer: Last comment: Should this be a H2? It breaks the usual page template. What might work is:

* move this para up to the bottom of the `## Description` section saying 'there are two examples that demonstrate percentage gap values with explicit and implicit container size below' linking to:
* Rename the example H3s to "Percentage gap value and explicit container size" and "Percentage gap value and implicit container size"

What do you reckon?

> Gap acts like an empty item/track. In other words instead of gap you could simply use an empty item/track in the container. The item/track's size would be the size set in `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.

Reviewer: Removing "simply":

```suggestion
A gap acts like an empty item or track. Instead of the gap, you could use an empty item or track in the container, and the item or track's size would be the size set in the `gap` property value. Percentage values are always calculated against the [content box](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#parts_of_a_box) size of the container element.
```

### Explicit container size

> If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 24x24px children. Both containers are explicitly set to 200px using `height: 200px` and the gap is set with `gap: 12.5% 0`.

Reviewer: ```suggestion
If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 20x20px children. Both containers are explicitly set to 200px high using `height: 200px` and the gap is set with `gap: 12.5% 0`.
```

> If the container has fixed size set then percentage value calculations are simple. Percentages are simply calculated from the sizes set on the container. Thus gap behaviour is consistant across all layouts. In the following example there are two containers one with grid layout and other with flex layout. The containers have five red 20x20px children. Both the containers have height explicitly set to 200px using `height: 200px`. And the gap is set using `gap: 10% 0`.

Reviewer: Minor reword:

```suggestion
If the container has a fixed size set, then gap percentage value calculations are based on the size of the container. Thus, gap behavior is consistent across all layouts. In the following example, there are two containers, one with a grid layout and the other with a flex layout. The containers have five red 20x20px children. Both containers are explicitly set to 200px using `height: 200px` and the gap is set with `gap: 10% 0`.
```

Note we're saying it's consistent but the spacing of items looks inconsistent. We should change the example so there's an equal number of gaps in each layout or explain why this may unexpectedly happen.

```html
<span>Grid</span>
<div id="grid">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
</div>
<span>Flex</span>
<div id="flex">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
</div>
```

```css hidden
body > div {
  background-color: #ccc;
  width: 200px;
  flex-flow: column;
}
```

```css
#grid {
  display: inline-grid;
  height: 200px;
  gap: 10% 0;
}

#flex {
  display: inline-flex;
  height: 200px;
  gap: 10% 0;
}

#grid > div,
#flex > div {
  background-color: coral;
  width: 20px;

>   height: 20px;

Reviewer: Note that we have five gaps in the flex layout and four in the grid layout (possibly margin collapse):

![image](https://github.com/mdn/content/assets/43580235/2165ef7e-73ce-407f-b167-573254694387)

we could do something like this with `min-height` so boxes are 24px

https://codepen.io/bsmth/pen/qBGYwyZ

In the current example, the gap is the same but the layouts look different. I'm not sure if that's intentional. What do you think?

Author: > Note that we have five gaps in the flex layout and four in the grid layout (possibly margin collapse):

Interesting, it's not margin collapse though. In case of grid the height gets divided equally among 5 [grid-rows](https://developer.mozilla.org/en-US/docs/Glossary/Grid_Row) then gaps *in between* the rows are subtracted from the row heights:
```
total height: 200px
grid rows without gaps: 5 * 40 = 200px
4 10% gaps in between rows + rows = 80 + 120 = 4 * 20 + (5 * 24)
```
From above calculations grid rows are of 24px, but the items are only 20px. The remaining 4px worth of grid area is remains blank:
![1](https://github.com/mdn/content/assets/87750369/1b2c0d26-504f-4955-97e4-333832e16f5b)
In above screenshot, taken from chrome, it's clear that grid rows are bigger than the grid items. Empty grid area has been pointed.

In case of flex, biggest flex-item on the row decides row height:
![2](https://github.com/mdn/content/assets/87750369/d92ecdd0-6ecb-49a1-b4f8-987ed271645a)
```
total height: 200px
flex rows + remaining empty area:  100 + 100 = 5 * 20 + 100
flex rows + grid gaps + remaining area: 100 + 80 + 20 = 5 * 20 + 4 * 20 + 20
```
In short, in case of our flex box, the items are stacked upwards and some area is left behind.

> In the current example, the gap is the same but the layouts look different. I'm not sure if that's intentional. What do you think?

It's not intentional, but the grid and flex containers are different. But we shouldn't get into such details because this is demo about gap. So we should set item heights to 24px to make both containers look more similar.

}
```

{{EmbedLiveSample("Explicit container size", "auto", "200px")}}

Now inspect the grid and flex elements using [Inspector tab in Web Developer Tools](https://firefox-source-docs.mozilla.org/devtools-user/page_inspector/how_to/open_the_inspector/index.html). In order to see the actual gaps hover mouse over `<div id="grid">` and `<div id="flex">` tags in the inspector. You will notice that the gap is same in both the cases which is 20px.

### Implicit container size

If size is not explicitly set on the container, then the percentage gap behaves different in case of grid and flex layouts. In the following example the containers don't have height explicitly set.

```html hidden
<span>Grid</span>
<div id="grid">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
</div>
<span>Flex</span>
<div id="flex">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
</div>
```

```css hidden
body > div {
  background-color: #ccc;
  width: 200px;
}

#grid {
  display: inline-grid;
  gap: 10% 0;
}

#flex {
  display: inline-flex;
  gap: 10% 0;
  flex-flow: column;
}

#grid > div,
#flex > div {
  background-color: coral;

>   width: 20px;
>   height: 20px;

Author: ```suggestion
  width: 24px;
  height: 24px;
```

}
```

{{EmbedLiveSample("Implicit container size", "auto", "200px")}}

> In case of the grid layout, percentage gap doesn't contribute to actual height of the grid. The container's height is calculated using `0px` gap, so the actual height turns out to be 120px (24px x5). Then the actual percentage gap is calculated using the content box's height, the gap turns out to be `12px` (120px x 10%). The gap is applied just before rendering. Thus the grid remains 120px high but it overflows due to the percentage gap added later just befor rendering.

Author: ```suggestion
In case of the grid layout, percentage gap doesn't contribute to actual height of the grid. The container's height is calculated using `0px` gap, so the actual height turns out to be 120px (24px x 5). Then the actual percentage gap is calculated using the content box's height, the gap turns out to be `12px` (120px x 10%). The gap is applied just before rendering. Thus the grid remains 120px high but it overflows due to the percentage gap added later just befor rendering.
```
